File: hw07/collcount.fs
\ check_forth.fs
\ Millard A. Arnold
\ 2023-04-28
\ 
\ For CS 331 Spring 2023
\ Assignment 7, Exercise 2

: collcount { n -- c }
  0
  begin
  n 1 > while
    n 2 mod 0 = if
      n 2 / to n
    else
      n 3 * 1 + to n
    endif
    1 +
  repeat
;  

File: hw07/collcount.pl
% Millard A. ArN0ld V
% 2023-04-30
% 
% For CS 331 Spring 2023
% Solution to Assignment 7 Exercise 4

% collcount\2
% collcount(+n, ?c)

collcount(1, 0).
collcount(N0, C0) :-
  N0 > 1,
  N0 mod 2 =:= 0,
  N1 is N0 / 2,
  collcount(N1, C1),
  C0 is C1 + 1.
collcount(N0, C0) :-
  N0 > 1,
  N0 mod 2 =:= 1,
  N1 is (3 * N0 + 1),
  collcount(N1, C1),
  C0 is C1 +  1.